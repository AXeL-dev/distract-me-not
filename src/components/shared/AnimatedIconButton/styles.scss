.animated-icon-button {
  transition: transform 0.2s;
  &.scale-0 {
    transform: scale(0);
  }
}

// .grow {
//   transition: all .2s ease-in-out;
//   &:hover {
//     transform: scale(1.1);
//   }
// }
