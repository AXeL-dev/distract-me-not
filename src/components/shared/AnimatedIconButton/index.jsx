import React, { Component } from 'react';
import { debug } from 'helpers/debug';
import { IconButton } from 'components';
import { isBoolean } from 'lodash';
import './styles.scss';

export class AnimatedIconButton extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isVisible: isBoolean(this.props.isVisible) ? this.props.isVisible : true, // do not use || operator with boolean values that takes "true" by default
      className: '',
      icon: this.props.icon,
      iconColor: this.props.iconColor,
      hideOnClick: this.props.hideOnClick,
    };
  }

  componentDidUpdate(prevProps, prevState) {
    //debug.log({ props: this.props, prevProps: prevProps, state: this.state, prevState: prevState });
    if (
      this.props.isVisible !== prevProps.isVisible &&
      this.props.isVisible !== this.state.isVisible
    ) {
      debug.warn('isVisible prop has changed:', this.props.isVisible);
      if (this.props.isVisible) {
        this.show();
      } else {
        this.hide();
      }
    }
  }

  setVisibilityProp = (value) => {
    if (this.props.onVisibilityChange) {
      this.props.onVisibilityChange(value);
    }
  };

  show = () => {
    this.setState({
      ...this.state,
      isVisible: true,
      className: '', // reset class
      icon: this.props.icon, // reset icon
      iconColor: this.props.iconColor,
    });
    this.setVisibilityProp(true);
  };

  hide = () => {
    this.setState({ ...this.state, isVisible: false });
    this.setVisibilityProp(false);
  };

  playHideAnimation = () => {
    // hide animation
    setTimeout(() => {
      this.setState({ ...this.state, className: 'scale-0' }); // scale to 0
      setTimeout(() => {
        if (this.props.hideAnimationIcon) {
          // change icon + color
          this.setState({
            ...this.state,
            className: '',
            icon: this.props.hideAnimationIcon,
            iconColor: this.props.hideAnimationIconColor || this.props.iconColor,
          });
          setTimeout(() => {
            this.setState({ ...this.state, className: 'scale-0' }); // rescale to 0
            setTimeout(() => {
              this.hide();
            }, 200);
          }, 500);
        } else {
          this.hide();
        }
      }, 200);
    }, 100);
  };

  handleClick = async (event) => {
    let shouldHide = this.state.hideOnClick;
    if (this.props.onClick) {
      shouldHide = await this.props.onClick(event);
    }
    if (shouldHide) {
      this.playHideAnimation();
    }
  };

  render() {
    return (
      this.state.isVisible && (
        <IconButton
          className={[
            'animated-icon-button',
            this.props.className || '',
            this.state.className || '',
          ].join(' ')}
          appearance={this.props.appearance || 'minimal'}
          icon={this.state.icon}
          iconColor={this.state.iconColor}
          iconSize={this.props.iconSize}
          tooltip={this.props.tooltip}
          tooltipPosition={this.props.tooltipPosition}
          onClick={this.handleClick}
        />
      )
    );
  }
}
