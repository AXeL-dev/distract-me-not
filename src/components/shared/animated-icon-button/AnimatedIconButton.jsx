import { Component, Fragment } from 'react';
import { Tooltip, IconButton } from 'evergreen-ui';
import { debug } from '../../../helpers/debug';
import './AnimatedIconButton.scss';

export default class AnimatedIconButton extends Component {

  constructor(props) {
    super(props);
    this.state = {
      isVisible: this.props.isVisible ? this.props.isVisible : true, // do not use || operator with boolean values that takes "true" by default
      className: '',
      icon: this.props.icon,
      hideOnClick: this.props.hideOnClick,
    };
  }

  componentDidUpdate(prevProps, prevState) {
    //debug.log({ props: this.props, prevProps: prevProps, state: this.state, prevState: prevState });
    if (this.props.isVisible !== prevProps.isVisible && this.props.isVisible !== this.state.isVisible) {
      debug.warn('isVisible prop has changed:', this.props.isVisible);
      if (this.props.isVisible) {
        this.show();
      } else {
        this.hide();
      }
    }
  }

  setVisibilityProp = (value) => {
    if (this.props.onVisibilityChange) {
      this.props.onVisibilityChange(value);
    }
  }

  show = () => {
    this.setState({
      ...this.state,
      isVisible: true,
      className: '', // reset class
      icon: this.props.icon, // reset icon
    });
    this.setVisibilityProp(true);
  }

  hide = () => {
    this.setState({ ...this.state, isVisible: false });
    this.setVisibilityProp(false);
  }

  playHideAnimation = () => {
    // hide animation
    setTimeout(() => {
      this.setState({ ...this.state, className: 'scale-0' }); // scale to 0
      setTimeout(() => {
        if (this.props.hideAnimationIcon) {
          this.setState({ ...this.state, className: '', icon: this.props.hideAnimationIcon }); // change icon
          setTimeout(() => {
            this.setState({ ...this.state, className: 'scale-0' }); // rescale to 0
            setTimeout(() => {
              this.hide();
            }, 200);
          }, 500);
        } else {
          this.hide();
        }
      }, 200);
    }, 100);
  }

  handleClick = (event) => {
    if (this.props.onClick) {
      this.props.onClick(event);
      if (this.state.hideOnClick) {
        this.playHideAnimation();
      }
    }
  }

  render() {
    return (
      <Fragment>
        {this.state.isVisible &&
          <Tooltip content={this.props.tooltip} position={this.props.tooltipPosition}>
            <IconButton
              className={`animated-icon-button ${this.props.className} ${this.state.className}`}
              appearance="minimal"
              icon={this.state.icon}
              iconSize={this.props.iconSize || 22}
              onClick={this.handleClick}
            />
          </Tooltip>
        }
      </Fragment>
    );
  }
}
