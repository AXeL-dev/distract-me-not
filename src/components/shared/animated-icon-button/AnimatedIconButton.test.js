import { render, fireEvent, waitFor } from "@testing-library/react";
import { CrossIcon } from "evergreen-ui";
import AnimatedIconButton from "./AnimatedIconButton";

it('renders correctly', () => {
  const { asFragment } = render(<AnimatedIconButton icon={CrossIcon} />);
  expect(asFragment()).toMatchSnapshot();
});

it('handles click', () => {
  const handleClick = jest.fn();
  const { container } = render(<AnimatedIconButton icon={CrossIcon} onClick={handleClick} />);
  fireEvent.click(container.firstChild);
  expect(handleClick).toHaveBeenCalledTimes(1);
});

it('hides on click when hideOnClick prop is set', async () => {
  const { container, getByRole } = render(<AnimatedIconButton icon={CrossIcon} hideOnClick={true} />);
  const button = getByRole('button'); // or container.firstChild (but it will not work once the button get removed after click)
  fireEvent.click(button);
  await waitFor(() => expect(button).not.toBeInTheDocument(), { timeout: 2000 });
});
