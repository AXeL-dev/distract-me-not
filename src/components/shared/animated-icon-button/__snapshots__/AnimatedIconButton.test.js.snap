// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <button
    class="css-1ii3p2c icon-button animated-icon-button   ub-fnt-fam_b77syt ub-mt_0px ub-fnt-sze_12px ub-f-wght_500 ub-ln-ht_32px ub-ltr-spc_0 ub-btrr_3px ub-bbrr_3px ub-btlr_3px ub-bblr_3px ub-pt_0px ub-pb_0px ub-pr_0px ub-pl_0px ub-ml_0px ub-mr_0px ub-mb_0px ub-h_32px ub-pst_relative ub-dspl_flex ub-algn-itms_center ub-flx-wrap_nowrap ub-just-cnt_center ub-w_32px ub-box-szg_border-box"
  >
    <div
      class="ub-dspl_inline-flex ub-box-szg_border-box"
    >
      <svg
        class="ub-w_22px ub-h_22px ub-box-szg_border-box"
        data-icon="cross"
        style="fill: currentColor;"
        viewBox="0 0 20 20"
      >
        <path
          d="M11.41 10l4.29-4.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L10 8.59l-4.29-4.3a1.003 1.003 0 00-1.42 1.42L8.59 10 4.3 14.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4.29-4.3 4.29 4.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L11.41 10z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </button>
</DocumentFragment>
`;
